### PR TITLE
reformat code, added comments, fixed some of the duplicated codes

### PR DESCRIPTION
**Pull request by ChinHongTan**
code is too lengthy, some parts of the code has too many nested for and if, making it difficult to read
lack of comments making it hard to figure out what is happening
make use of spaces to increase readability
use loops instead of copy & paste can help reduce file size and increase maintainability
avoid lots of if elses by using early returns and/or switch to simplify the code

### CODE REVIEW

File: cmds/minigames/connect4.js [hunk 1]
```diff
@@ -105,55 +166,110 @@ class connect4 extends two_p_Game{
...
                     }
 
-                    console.log("y1")
-                    if(user.id === game.yellowid && current_color !== "yellow") return
-                    console.log("y2")
-
-                    switch(game.place("1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣🛑".indexOf(reaction.emoji.name)/3,current_color)){
-
+                    console.log("y1");
+                    if (user.id === game.yellowid && current_color !== "yellow")
```
Comment (xiaomiao-fong) [inline cmds/minigames/connect4.js:235]: Codacy has a fix for the issue: [Expected { after 'if' condition.](https://app.codacy.com/gh/xiaomiao-fong/Chen_bot/pullRequest?prid=7807535)
```suggestion
                    if (user.id === game.yellowid && current_color !== "yellow")
```

File: cmds/minigames/connect4.js [hunk 2]
```diff
@@ -1,86 +1,147 @@
...
                     }
 
-                    console.log("r1")
-                    if(user.id === game.redid && current_color !== "red") return
-                    console.log("r2")
-
-                    switch(game.place("1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣🛑".indexOf(reaction.emoji.name)/3,current_color)){
-
+                    console.log("r1");
+                    if (user.id === game.redid && current_color !== "red")
```
Comment (xiaomiao-fong) [inline cmds/minigames/connect4.js:107]: Codacy has a fix for the issue: [Expected { after 'if' condition.](https://app.codacy.com/gh/xiaomiao-fong/Chen_bot/pullRequest?prid=7807535)
```suggestion
                    if (user.id === game.redid && current_color !== "red")
```

File: cmds/minigames/connect4.js [hunk 3]
```diff
@@ -1,84 +1,141 @@
...
```
Comment (xiaomiao-fong) [inline cmds/minigames/connect4.js:99]: Codacy has a fix for the issue: [Expected { after 'if' condition.](https://app.codacy.com/gh/xiaomiao-fong/Chen_bot/pullRequest?prid=7807535)
```suggestion
                    if (user.id === game.redid && current_color !== "red")
```

File: cmds/minigames/connect4.js [hunk 4]
```diff
@@ -175,288 +289,320 @@ class connect4 extends two_p_Game{
...
-            if(board[y][x] === "⚪"){
-                return {count : count, live : 1}
+    search2(x, y, color, dx, dy, count, board) {
+        if (x < 0 || x > 6 || y < 0 || y > 5) return { count: count, live: 0 };
+        if (board[y][x] !== color) {
+            if (board[y][x] === "⚪") {
+                return { count: count, live: 1 };
             }
-            return {count : count, live : 0}
+            return { count: count, live: 0 };
```
Comment (xiaomiao-fong) [inline cmds/minigames/connect4.js:458]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/xiaomiao-fong/Chen_bot/pullRequest?prid=7807535)
```suggestion
            return { count, live: 0 };
```

File: cmds/minigames/connect4.js [hunk 5]
```diff
@@ -175,288 +289,320 @@ class connect4 extends two_p_Game{
...
     }
 
-    search2(x, y, color, dx, dy, count, board){
-
-        if(x<0 || x>6 || y<0 || y>5) return {count : count, live : 0}
-        if(board[y][x] !== color){
-            if(board[y][x] === "⚪"){
-                return {count : count, live : 1}
+    search2(x, y, color, dx, dy, count, board) {
+        if (x < 0 || x > 6 || y < 0 || y > 5) return { count: count, live: 0 };
```
Comment (xiaomiao-fong) [inline cmds/minigames/connect4.js:453]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/xiaomiao-fong/Chen_bot/pullRequest?prid=7807535)
```suggestion
        if (x < 0 || x > 6 || y < 0 || y > 5) return { count, live: 0 };
```

File: cmds/minigames/connect4.js [hunk 6]
```diff
@@ -175,288 +289,320 @@ class connect4 extends two_p_Game{
...
+    // calculate the best move
+    c4minimax(board, stack, depth, max, lx, ly) {
+        if (this.check(lx, ly, board) === 2 || depth === 0) {
+            let temp1 = this.board_evaluation(board, "🟡");
+            let temp1score =
+                temp1.two + (temp1.three * 5) / 3 + temp1.four * 15;
+            let temp2 = this.board_evaluation(board, "🔴");
+            let temp2score =
+                temp2.two + (temp2.three * 5) / 3 + temp2.four * 20;
+            if (temp1.four > 0 || temp2.four > 0)
```
Comment (xiaomiao-fong) [inline cmds/minigames/connect4.js:483]: Codacy has a fix for the issue: [Expected { after 'if' condition.](https://app.codacy.com/gh/xiaomiao-fong/Chen_bot/pullRequest?prid=7807535)
```suggestion
            if (temp1.four > 0 || temp2.four > 0)
```

File: cmds/minigames/connect4.js [hunk 7]
```diff
@@ -102,55 +159,112 @@ class connect4 extends two_p_Game{
...
                     }
 
-                    console.log("y1")
-                    if(user.id === game.yellowid && current_color !== "yellow") return
-                    console.log("y2")
-
-                    switch(game.place("1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣🛑".indexOf(reaction.emoji.name)/3,current_color)){
-
+                    console.log("y1");
+                    if (user.id === game.yellowid && current_color !== "yellow")
```
Comment (xiaomiao-fong) [inline cmds/minigames/connect4.js:229]: Codacy has a fix for the issue: [Expected { after 'if' condition.](https://app.codacy.com/gh/xiaomiao-fong/Chen_bot/pullRequest?prid=7807535)
```suggestion
                    if (user.id === game.yellowid && current_color !== "yellow")
```

File: cmds/minigames/connect4.js [hunk 8]
```diff
@@ -175,288 +289,320 @@ class connect4 extends two_p_Game{
...
-
-    search(x, y, color, dx, dy, count, board = this.map){
-
-        if(x<0 || x>6 || y<0 || y>5) return count
-        if(color !== board[y][x]) return count
-
-        return this.search(x+dx, y+dy, color, dx, dy, count+1, board)
+    // (not sure what this whole code does)
+    search(x, y, color, dx, dy, count, board = this.map) {
+        if (x < 0 || x > 6 || y < 0 || y > 5) return count;
```
Comment (xiaomiao-fong) [inline cmds/minigames/connect4.js:446]: Codacy has a fix for the issue: [Expected { after 'if' condition.](https://app.codacy.com/gh/xiaomiao-fong/Chen_bot/pullRequest?prid=7807535)
```suggestion
        if (x < 0 || x > 6 || y < 0 || y > 5) {return count;}
```

File: cmds/minigames/connect4.js [hunk 9]
```diff
@@ -1,84 +1,141 @@
...
-                let field1 = {name: "Now : ",value: msg.author.username + client.language.commands.game[userlang].ones_turn , inline: true}
-                let rmsg = await msg.author.send(client.EmbedMaker(msg, game.stringify(), client.colors.red,[field1]))
-                let ymsg = await iuser.send(client.EmbedMaker(msg, game.stringify(), client.colors.red,[field1]))
-
-                for(var i = 0; i < 8; i++){
-                    rmsg.react(game.reactions[i])
-                    ymsg.react(game.reactions[i])
+                let field1 = {
+                    name: "Now : ",
+                    value: msg.author.username + client.language.commands.game[userlang].ones_turn,
```
Comment (ChinHongTan) [inline cmds/minigames/connect4.js:33]: Language options like this is very lengthy.
Consider using 
```suggestion
                    let language = client.language.commands.game[userlang]
                    // and then in your code
                    value: msg.author.username + language.ones_turn,
```

File: cmds/minigames/connect4.js [hunk 10]
```diff
@@ -1,84 +1,141 @@
...
-
+                const redcollector = rmsg.createReactionCollector(filter, {
+                    idle: 300 * 1000,
+                });
+                const yellowcollector = ymsg.createReactionCollector(filter, {
+                    idle: 300 * 1000,
+                });
+
+                // merge 2 collector if possible
+                redcollector.on("collect", async (reaction, user) => {
```
Comment (ChinHongTan) [inline cmds/minigames/connect4.js:76]: the two collectors are essentially doing the same thing, consider using a function to merge the two collectors

File: cmds/minigames/connect4.js [hunk 11]
```diff
@@ -175,288 +289,320 @@ class connect4 extends two_p_Game{
...
-    board_evaluation(board, color){
-
-        let result = {two : 0, three : 0, four : 0}
-
-        for(var i = 0; i < 6; i++){
-            for(var j = 0; j < 7; j++){
-
-                if(board[i][j] === color){
+    // not sure what this does, but simplify it if possible
+    board_evaluation(board, color) {
```
Comment (ChinHongTan) [inline cmds/minigames/connect4.js:545]: Not sure what this does, but there is a large chunk of code block that looks identical, and complex nested for loop, if and switch statement
Simplify it if possible

File: cmds/minigames/connect4.js [hunk 12]
```diff
@@ -102,55 +159,112 @@ class connect4 extends two_p_Game{
...
+                                    msg,
+                                    game.stringify(),
+                                    client.colors.white
+                                )
+                            );
+                            await ymsg.edit(
+                                client.EmbedMaker(
+                                    msg,
+                                    game.stringify(),
+                                    // client.colors / white =============== I'm not sure what this "/" does so I assume it is a typo
```
Comment (ChinHongTan) [inline cmds/minigames/connect4.js]: I think this is a typo here
```suggestion
                                    // client.colors / white
                                    client.colors.white
```

**Review by ChinHongTan — COMMENTED** [review]
Added some suggestions

File: cmds/minigames/connect4.js [hunk 13]
```diff
@@ -102,55 +159,113 @@ class connect4 extends two_p_Game{
...
+                                    game.stringify(),
+                                    client.colors.white
+                                )
+                            );
+                            await ymsg.edit(
+                                client.EmbedMaker(
+                                    msg,
+                                    game.stringify(),
+                                    // client.colors / white
+                                    client.colors.white
```
Comment (ChinHongTan) [inline cmds/minigames/connect4.js:195]: duplicated line
```suggestion                               
```

**Comment by ChinHongTan** [comment]
I am leaving this branch here
You can merge it if you like, or do further changes of your own

**Pull request merged**